### PR TITLE
Implement JWT authentication

### LINKED ISSUE #29: JWT Authentication
Implement authentication so that our backend knows that the requests it receives are legitimate

### PR DESCRIPTION
**Pull request by cheehongw**
Basic workflow for JWT token authentication

Closes #29 

1. Client logs in, and the server returns a JWT as a session token.
2. This token is stored on the client-side as an HTTP-only cookie (for security reasons).
3. For subsequent requests, the browser automatically attaches this cookie.
4. On the server-side, the JWT from the cookie is verified to authenticate the user.

### CODE REVIEW

**Comment by cheehongw** [comment]
Ready for review

**Review by Eclipse-Dominator — CHANGES_REQUESTED** [review]
Sorry for another review. found another bug!

Right now, for testing, we have used a dummy user in auth-slice. when setting it to null.

the following will happen for users that are logged in and access the page on new tab/refresh:
1. user in auth-slice is currently null
2. Page will redirect to /login
3. async method `getSessionUser` will fetch current active user
4. router will swap to use logged in route
5. /login page will redirect to '404' as /login is not routed in the new route

Thus, the following fix should be needed:
- Redirect /login to some other page (maybe the last page visited?)
- Some way to prevent page from redirecting to /login immediately
   - it's possible to fix this by persisting active user in memory
   - or wait for getsessionuser to finish before loading other dom element

**Comment by cheehongw** [comment]
Thanks for the review @Eclipse-Dominator. 
1. I'll standardize what is returned from the backend by `getSessionUser` and `login/register` by refactoring some stuff

2. As for logout button not invalidating the session token, an additional call will be made to the server for the server to invalidate the cookie.

3. Will remove the unused vars

4. Yes, I encountered that actually. Forgot to include in the PR. Current plan is to redirect the logged in user to `/`

**Review by Eclipse-Dominator — CHANGES_REQUESTED** [review]
Generally lgtm except for the logout using the wrong http protocol. 

Note the following interactions below:
- a logged in user open 2 tabs => logged out on 1 tab, other tab stays logged in until a interaction that requires interaction with user service again
- a user opens 2 login tab => log in 2 diff accounts 

I think caching user in `localstorage` might not be a bad idea for this. Generally, the issue is 
 effect of login/log out does not affect other opened tabs

**Comment by cheehongw** [comment]
For this interaction:

> a logged in user open 2 tabs => logged out on 1 tab, other tab stays logged in until a interaction that requires interaction with user service again

When the user logs out on tabX, the redux state is updated for tabX, resulting in a rerender for tabX but not tabY. 
We can look into approaches to ensure the same redux state for all tabs, so that a change in redux state caused by tabX would be sync to all the other redux states in tabY, tabZ, etc..  

-----------

For now, I think allowing two different user sessions at the same time might be more tricky than it is worth. 

While we can store the auth-session token in localStorage, this would require modifications on the frontend to handle the different tokens for different accounts. The current implementation does not require much changes to the frontend at all.

Maybe one work-around could be to use two separate browsers if we want to test two separate users? Or maybe incognito mode?

File: frontend/src/api/auth.ts [hunk 1]
```diff
@@ -9,11 +15,43 @@ export async function login(username: String, password: String) {
...
+    } catch (error) {
+        return response
+    }
+}
+
+/**
+ * Logs the user out and server sets the token to expire
+ */
+export async function logOut() {
+    const response = await userServiceClient.get('/logout');
```
[the remark this answers is not in the document]
Comment (cheehongw) [reply frontend/src/api/auth.ts:55]: changed the backend to use `GET` instead

**Review by Eclipse-Dominator — APPROVED** [review]
lgtm!

**Pull request merged**